pkg/disk/partition_table.go

```diff
@@ -1144,3 +1112,286 @@ func mkESP(size uint64, ptType PartitionTableType) (Partition, error) {
+func (options *CustomPartitionTableOptions) getfstype(fstype string) (string, error) {
+	if fstype != "" {
+		return fstype, nil
+	}
+
+	if options.DefaultFSType == FS_NONE {
+		return "", fmt.Errorf("no filesystem type defined and no default set")
+	}
+
+	return options.DefaultFSType.String(), nil
+}
+
+// NewCustomPartitionTable creates a partition table based almost entirely on the disk customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.DiskCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
+	if options == nil {
+		// init options with defaults
+		options = &CustomPartitionTableOptions{
+			PartitionTableType: PT_GPT,
+		}
+	}
+
+	if customizations == nil {
+		customizations = &blueprint.DiskCustomization{}
+	}
+
+	errPrefix := "error generating partition table:"
+
+	// validate the partitioning customizations before using them
+	if err := customizations.Validate(); err != nil {
+		return nil, fmt.Errorf("%s %w", errPrefix, err)
+	}
+
+	pt := &PartitionTable{}
+
+	// TODO: Handle partition table type in customizations
+	switch options.PartitionTableType {
+	case PT_GPT, PT_DOS:
+		pt.Type = options.PartitionTableType
+	case PT_NONE:
+		// default to "gpt"
+		pt.Type = PT_GPT
+	default:
+		return nil, fmt.Errorf("%s invalid partition table type enum value: %d", errPrefix, options.PartitionTableType)
+	}
+
+	// TODO: switch to ensure ESP in case customizations already include it
+	if err := AddPartitionsForBootMode(pt, options.BootMode); err != nil {
+		return nil, fmt.Errorf("%s %w", errPrefix, err)
+	}
+
+	// The boot type will be the default only if it's a supported filesystem
+	// type for /boot (ext4 or xfs). Otherwise, we default to xfs.
+	// FS_NONE also falls back to xfs.
+	var bootFsType FSType
+	switch options.DefaultFSType {
+	case FS_EXT4, FS_XFS:
+		bootFsType = options.DefaultFSType
+	default:
+		bootFsType = FS_XFS
+	}
+
+	if needsBoot(customizations) {
+		// we need a /boot partition to boot LVM or Btrfs, create boot
+		// partition if it does not already exist
+		if err := AddBootPartition(pt, bootFsType); err != nil {
+			return nil, fmt.Errorf("%s %w", errPrefix, err)
+		}
+	}
+
+	for _, part := range customizations.Partitions {
+		switch part.Type {
+		case "plain", "":
+			if err := addPlainPartition(pt, part, options); err != nil {
+				return nil, fmt.Errorf("%s %w", errPrefix, err)
+			}
+		case "lvm":
+			if err := addLVMPartition(pt, part, options); err != nil {
+				return nil, fmt.Errorf("%s %w", errPrefix, err)
+			}
+		case "btrfs":
+			addBtrfsPartition(pt, part)
+		default:
+			return nil, fmt.Errorf("%s invalid partition type: %s", errPrefix, part.Type)
+		}
+	}
+
+	if err := EnsureRootFilesystem(pt, options.DefaultFSType); err != nil {
+		return nil, fmt.Errorf("%s %w", errPrefix, err)
+	}
+
+	if len(options.RequiredMinSizes) != 0 {
+		pt.EnsureDirectorySizes(options.RequiredMinSizes)
+	}
+
+	pt.relayout(customizations.MinSize)
```

Comment:
AFAICT we have no test for this value, yes? If I do:
```go
diff --git a/pkg/disk/partition_table.go b/pkg/disk/partition_table.go
index 9bed8eb61..213414340 100644
--- a/pkg/disk/partition_table.go
+++ b/pkg/disk/partition_table.go
@@ -1237,7 +1237,7 @@ func NewCustomPartitionTable(customizations *blueprint.DiskCustomization, option
                pt.EnsureDirectorySizes(options.RequiredMinSizes)
        }
 
-       pt.relayout(customizations.MinSize)
+       pt.relayout(0)
        pt.GenerateUUIDs(rng)
 
        return pt, nil

```
no test fails - should be easy enough to add though, the testing is so nice already which makes extending trivial